### ISSUE TITLE
Generated defaultValues are not applied after first render

### ISSUE THREAD

**Issue by stefanv** [react-hook-form/react-hook-form#1558]
I have a form that includes controlled components (several Material-UI checkboxes).  By specifying `defaultValues`, I am able to reset the form successfully.  However, if `defaultValues` changes depending on props, the reset of the controlled components fails.

E.g., this works:

```
const MyForm = ({ acls }) => {
  ...
  const defaultValues = {
    name: '',
    acls: [false, false, false, false, false, false, false]
  };
  const { handleSubmit, register, errors, reset, setValue, control } = useForm({
    defaultValues
  });
  ...
}
```

This does not:

```
const MyForm = ({ acls }) => {
  ...
  const defaultValues = {
    name: '',
    acls: Array(acls.length).fill(false)
  };
  const { handleSubmit, register, errors, reset, setValue, control } = useForm({
    defaultValues
  });
  ...
}
```

I can see that my top-level component gets re-rendered, as `acls` changes along the way from 0 to 7 elements.  Thus, in the first example, `defaultValues` is constant, while in the second it is modified because of the props updating.  Is it possible that, somehow, useForm is not being executed upon the second render?

My apologies, I am unsure how to build a codesandbox that replicates this behavior.  If you have a hint on how to replicate the props update there, I'd be happy to give it a shot.

**Expected behavior**

When I recreate a form using `useForm` with different parameters, those parameters should take effect.

**Desktop (please complete the following information):**
 - OS: Linux
 - Browser: Firefox
 - Version: 73
 - react-hook-form: 5.6.1

Thank you very much for react-hook-form.  I really like what I'm seeing so far!

**Comment by devrnt** [comment]
You should use `defaultValues` only for initialization, so when you invoke `useForm`. When you want to update the values in the _future_ you should use the `reset` or `setValue`api. 

I think you can use a `useEffect` where you invoke `reset({ acls: Array(acls.length).fill(false) })` instead of trying to update the `defaultValues`.

**Comment by stefanv** [comment]
This is a functional component, so I *am* providing `defaultValues` to `useForm`. Surely, you're not meant to keep track of whether `useForm` is only executed on a single render?

Clarifying: I am *not* trying to update `defaultValues`. It is calculated when the component is rendered.

**Comment by bluebill1049** [comment]
where is the codesandbox to reproduce?

**Comment by stefanv** [comment]
I will try and figure out how to build a redux sandbox.

With redux, it is quite typical for props to change, and of the defaultValues of the form depends on those props, this type of problem will occur.

The problem with reset is that you cannot simply call it in a functional container, otherwise it will lead to continual re-rendering.

**Comment by bluebill1049** [comment]
Why would redux pass new props while users interacting with the form. It’s not redux form where we store form state inside the redux. If value arrive asynchronous then use setValue, if entire form default values need to be updated then use reset. I don’t see conflicts between this lib and redux.

**Comment by stefanv** [comment]
Ok, I think I see: this would need to be done with `useEffect` probably? I'll give that a shot and report back. Thanks!

**Comment by bluebill1049** [comment]
yea, if you got an issue to send that codesandbox, i am more than happy to assist.

**Comment by stefanv** [comment]
@bluebill1049 I managed to build a sandbox that replicates the behavior I see at https://codesandbox.io/s/react-hook-form-controller-vhuqg?file=/src/index.js

The `useEffect` workaround didn't do the trick, but most likely I'm missing something obvious.

**Comment by bluebill1049** [comment]
hey @stefanv, I find your example hard to follow, can you explain the steps to reproduce the problem?

**Comment by stefanv** [comment]
Sure! The form get dynamically updated depending on the prop nr. So, if you increase the counter, the number of checkboxes increase correspondingly.

However, defaultState is cached, and does not get updated, or so it seems.

To see it in action, increase the number of checkboxes, check them all, and try to reset the form. Only the first box resets.

**Comment by bluebill1049** [comment]
@stefanv you want to take a look at my example: https://codesandbox.io/s/determined-tree-d4yyc?file=/src/index.js

**Comment by stefanv** [comment]
I'm wondering whether this would work with controlled components, and when we don't have nr as a field in the form (which I typically don't). I will convert your example to use controlled components and verify.

**Comment by stefanv** [comment]
The moment I change those inputs to controlled components, the proposed solution no longer works.  That was the reason for using `defaultValues` in the first place, since the docs state: 

> Note: You will need to supply defaultValues at useForm to reset Controller components' value.

**Comment by bluebill1049** [comment]
https://codesandbox.io/s/determined-tree-d4yyc?file=/src/index.js what about this?

**Comment by stefanv** [comment]
@bluebill1049 When changing the input to a material-UI checkbox, it does not reset properly:

https://codesandbox.io/s/loving-fire-5xv98?file=/src/index.js

**Comment by bluebill1049** [comment]
https://codesandbox.io/s/fervent-solomon-2u8cr?file=/src/index.js

**Comment by stefanv** [comment]
@bluebill1049 That works because you are manually setting the two boxes to false, but if you grow the form that doesn't work any more.

**Comment by bluebill1049** [comment]
if that's your use case, then you should have used `useFieldArray`.

**Comment by bluebill1049** [comment]
it's late here, i will have a play with your example tmr again.

**Comment by bluebill1049** [comment]
https://codesandbox.io/s/fervent-solomon-2u8cr?file=/src/index.js:637-642 last attempt before leaving it for tmr.

**Comment by stefanv** [comment]
Thanks for giving it another shot, especially so late!  This one looks promising, and hopefully we get an example for the documentation from the experiment.

**Comment by stefanv** [comment]
OK, finally, I think we have it!

https://codesandbox.io/s/react-hook-form-controller-9zxy3?file=/src/index.js

Importantly, the `Controller` needed a `defaultValue`, and no more `defaultValues` is needed.

**Comment by bluebill1049** [comment]
awesome, i would probably have used useFieldArray for dynamic fields.

**Comment by stefanv** [comment]
I'm including the full example here for posterity, in case the code sandbox goes away.  The reason I did not use `useFieldArray` is because, in my actual example, I had a more complex form to set up.

```javascript
import React, { useState, useEffect } from "react";
import ReactDOM from "react-dom";
import { useForm, Controller } from "react-hook-form";

import { Checkbox } from "@material-ui/core";

const Form = ({ nr }) => {
  nr = parseInt(nr, 10);

  const { handleSubmit, reset, control } = useForm();

  const onSubmit = data => {
    console.log("Submit with:", data);
    reset();
  };

  useEffect(() => {
    reset({
      checkbox: Array(nr).fill(false)
    });
  }, [reset, nr]);

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      {Array.from(Array(nr).keys()).map(x => (
        <div>
          Item {x}
          <Controller
            as={Checkbox}
            key={x}
            name={`checkbox[${x}]`}
            control={control}
            defaultValue={false}
          />
        </div>
      ))}
      <button type="submit">Submit & reset</button>
    </form>
  );
};

const VariableForm = () => {
  let [nr, setNr] = useState("1");

  const onInput = event => {
    setNr(parseInt(event.target.value, 10));
  };

  return (
    <div>
      <p>Increase the counter below to build a larger form.</p>

      <input type="number" name="count" defaultValue="1" onInput={onInput} />
      <Form nr={nr} />
    </div>
  );
};

function App() {
  return VariableForm();
}

const rootElement = document.getElementById("root");
ReactDOM.render(<App />, rootElement);
```

**Comment by bluebill1049** [comment]
Thanks for sharing solution @stefanv

**Comment by hiimnhan** [comment]
> awesome, i would probably have used useFieldArray for dynamic fields.

Please create a codesandbox with useFieldArray, I have stucked at this. Many thanks!
